[Kafka Streams] Clean-up script [WIP]

mjsax opened the pull request:
Anyof: @guozhangwang @enothereska @dguy @miguno 

Two variants:
1. python client + bash script (seems not to be the right way, due to external dependencies)
2. java only (bash script to call the java program not included yet)

mjsax committed:
Initial code drop for clean-up script

mjsax committed:
added Java CleanUp-Client

dguy commented on streams/src/main/java/org/apache/kafka/streams/tools/KafkaStreamsCleanupClient.java:
```diff
@@ -0,0 +1,249 @@
+    private static OptionSpec<String> intermediate_topics_option;
+    private static OptionSpec<String> operator_option;
+    private static OptionSpec<String> window_option;
+    private static OptionSpec<String> state_dir_option;
```
You might want to change the names of these to be standard Java naming, i.e, camel case. I'd be surprised if Checksytle doesn't complain about these.

dguy commented on streams/src/main/java/org/apache/kafka/streams/tools/KafkaStreamsCleanupClient.java:
```diff
@@ -0,0 +1,249 @@
+                }
+            }
+
+            // TODO fix -- simply sleep is not good enough
```
I believe we can do this by submitting OffsetRequests for the topics, i.e., get the earliest and latest offsets for each partition and when the difference is 0 then we know we have 0 messages in the partition

mjsax committed:
checkstyle

enothereska commented on build.gradle:
```diff
@@ -672,6 +672,7 @@ project(':streams') {
   archivesBaseName = "kafka-streams"
 
   dependencies {
+    compile project(':core')
```
@mjsax I'm afraid we are not supposed to introduce a Kafka Core dependency on kafka streams. One option would be to put this tool into the kafka tools package, instead of the Kafka Streams package. @ijuma what do you think? I don't see any Kafka Streams dependencies here.

ijuma replied:
@enothereska that's probably the easiest solution. Worth checking what @guozhangwang thinks.

mjsax replied:
@guozhangwang Any comment on this? If not, I will leave as is.

ijuma replied:
A dependency on `core` is definitely not OK (`core` actually means `broker and old clients` for historical reasons). However, it would be good to know what @guozhangwang thinks is the best solution.

guozhangwang replied:
@enothereska and @ijuma is right. We are trying deliberately to not include `core` as a dependency of `streams`.

One example is for the Zkclient-related module for auto-creating topics, before `KIP-4` was introduced. We intentionally duplicate some of the code from `core` in order to not introducing the dependency.

I'm OK with putting it under the kafka tools package, it should be fine.

ijuma commented on kafka-streams-cleanup.py:
```diff
@@ -0,0 +1,163 @@
+import subprocess
+import shutil
+
+def print_usage (appname):
```
Why are we including this Python script? It seems out of place (it requires a bunch of things that are not part of Kafka).

mjsax replied:
We just tried out different possibilities. The Python script will most likely get removed and we will use the Java Client.

guozhangwang commented:
Copying some of my comments in the reprocessing wiki page here:

```
One problem with the "cleanup" API proposed in KAFKA-3185 is that, step 1) delete internal kafka topics / 2) delete application committed offsets / 3) cleanup user topics, as listed in manual cleanup would better be triggered only once on one of the instances, whereas step 4): local state store data can be cleaned up separately on each instance. But requiring users to specify all the internal topic names like we did in the script is a bit too much for users.

And another thing is that, regarding step 3), it is not clear in the auto clean-up scenario that it should be handled by Kafka Streams or users should worry about those themselves. More specifically, delete-then-recreate may not be the proper approach for all use cases.

So thinking it a bit more, now I'm thinking maybe we can do sth. like the following:
Make KafkaStreams.cleanup() only do the local cleanup (i.e. step 4) by deleting everything under the application id sub-directory, which is required to be called only after KafkaStreams.close().
Add a new admin tool for cleaning up the application-related Kafka topics as well as their offsets, given the specific application id, but no need the topic names, since we can list all topics in Kafka, and delete those with application id as prefix and -changelog and -repartition as suffix; this may have some risks of deleting non-application-owned topics but should work perfectly for most of the cases; and in the future when we introduce hierarchical topics (https://cwiki.apache.org/confluence/display/KAFKA/Hierarchical+Topics) it will be completely safe to do.
Then use delete-offset request to delete topics admin as in KAFKA-3059 to delete the offsets for all user / internal topics as well.
The usage pattern then is that, users first call KS.close() on all instances, and then KS.cleanup() on all instances, and then trigger this tool for step 1) and 2), and then handle their own topics as per their own needs, truncating or deleting or anything else. This pattern can be encoded in users' source manager or their own stream job management tooling (like in C3, for example).
```

@mjsax What do you think about the above proposal? I'm not suggesting we should necessarily considering doing the above pattern in this script, as the current approach is supposed only for development cycles and hence requiring it to be called on each instance would be OK because most likely there will be only one instance. But moving forward we could benefit from only requiring to call the admin tool only once.

miguno commented on streams/src/main/java/org/apache/kafka/streams/tools/KafkaStreamsCleanupClient.java:
```diff
@@ -0,0 +1,249 @@
+
+        List<String> allTopics = scala.collection.JavaConversions.seqAsJavaList(zkUtils.getAllTopics());
+
+        resetSourceTopicOffsets();
```
+1 to a readable "overview" in code that explains what this class does. ;-)

miguno commented on streams/src/main/java/org/apache/kafka/streams/tools/KafkaStreamsCleanupClient.java:
```diff
@@ -0,0 +1,249 @@
+        System.out.println("Deleted " + stateStore.getAbsolutePath());
+    }
+
+    private void deleteDirectory(File directory) {
```
FWIW: Kafka has utility methods for removing dirs under `kafka.utils.CoreUtils.delete()` but IIRC that's part of the `core` dependency that we don't want to include...

ijuma replied:
Also in `org.apache.kafka.common.utils.Utils` (`delete(File file)`).

mjsax committed:
removed python stuff
added sript to call StreamsCleanupClient from command line
fixed issus "wait until changelog got cleared"

mjsax committed:
fixed window changelog topic names

mjsax commented:
Closed in favor of https://github.com/apache/kafka/pull/1636
